File: scripts/coverage-prod.py
#!/usr/bin/env python3
"""COVERAGE-PROD - the production-slice coverage floor.

`cargo llvm-cov --fail-under-lines 80` is dominated by test-module lines: the total
floor does not fire until production coverage falls below ~44%. This checker reads
the JSON export `make coverage` writes alongside that run and enforces a second,
production-only floor.

A production line is one NOT inside the brace extent of any `#[cfg(test)]` item. This
is deliberately NOT "everything above the file's first line-anchored `#[cfg(test)]`"
(the `prod()` cut `READONLY-UI`, `NOBLOCK`, and `WIRED` use) - that cut misclassifies
6,281 production lines tree-wide because src/changes.rs, src/cli.rs, and src/lib.rs
each hold more than one `#[cfg(test)]` item. See
openspec/changes/gate-integrity/design.md -> Decision 1a.

Line-counting rule (stated because two exist, per Decision 1a): a line is
"instrumented" when at least one of its segments carries `hasCount`, and "covered"
when the maximum count among those segments exceeds zero. This is the `hasCount`
rule, not cargo-llvm-cov's own per-function `totals`, which yields a different
denominator for the same report.

Known limit, on the same terms NOSPAWN-GREP and WIRED record theirs: the brace
counter used to find a #[cfg(test)] item's extent does not parse Rust, only mask
comments and string/char literal content (so an unbalanced brace inside a string
cannot desync it - see mask_non_code below). The number of extents found is printed
per file specifically so a tracker defeated in some other way shows up as a
suspicious count rather than a silently shifted floor.

usage: python3 coverage-prod.py <llvm-cov-json-report> [<degraded-coverage-map>]
environment:
    PROD_MIN                overrides the floor (a percentage, e.g. "80"). Defaults
                             to the constant below, which SHALL be this checker's
                             own measured production-slice figure on the
                             unmodified tree, rounded down to a whole point - not
                             a number transcribed from a planning document.

gate-integrity task 6.4: this checker also requires every line of every
`covers` range in a degraded-coverage map to be executed - a `proof` that
resolves and a `covers` range that resolves (tests/degraded_coverage.rs's own
job) still leave open whether the degraded path actually RAN, which only a
real per-line coverage report can answer. The map path is the checker's
SECOND positional argument, not an environment variable and not a default:
the covers-range check runs only when a map is named on the command line, so
the enforcing invocation (`make coverage`, which always names
tests/degraded-coverage.toml) opts the check IN rather than every other
caller having to opt it OUT through an escape-hatch environment variable -
gate-integrity's own correction, task 6.4b, removed the environment-variable
skip flag this check used to carry, as a waiver inside the guard whose whole
thesis is that waivers are the defect. `tests/coverage_prod.rs`'s
production-floor fixtures accordingly pass no second argument at all; its
covers-range scenarios pass a small fixture map as that argument instead of
the real one.

The map is read with a small, regex-based extractor (`parse_covers_map`
below) rather than a TOML parser, on `scripts/gates/gate-mech1.py`'s own
established terms for a narrow, fully-controlled, checked-in input shape -
the file is never third-party input, and pulling in a TOML dependency
(stdlib `tomllib` needs Python 3.11, which `AGENTS.md` does not otherwise
require of this repository) to parse six fixed keys would be new machinery
for a fact this script can read directly. Known limit, recorded on
`mask_non_code`'s own terms: the extractor does not parse Rust either, so a
`covers` range is checked whether or not it is actually within the
array-of-tables structure the coverage spec assumes - it is checked against
the named map file itself, which `tests/degraded_coverage.rs` already keeps
honest for the real tests/degraded-coverage.toml.
"""
import json
import os
import re
import sys
from pathlib import Path

# Task 1.3: this SHALL be the checker's own output on the unmodified tree, rounded
# down to a whole point. See openspec/changes/gate-integrity/design.md -> Decision 1a
# and Risks ("PROD_MIN is set from the checker's own output at implementation time").
#
# Measured at implementation time (task 1.3), from this checker's own hasCount rule
# against a real `cargo llvm-cov --json` export of the unmodified tree (25 files under
# src/, 22,285 hasCount-instrumented lines total, matching design.md's own total under
# the same rule): production 96.07% (3544/3689 lines), test-module 96.43%
# (17933/18596). This differs from design.md's own 97.28% estimate (6623/6808) because
# the tree has grown since that estimate was written; per Decision 1a, the figure
# actually produced here - not the design document's - is what SHALL be recorded.
DEFAULT_PROD_MIN = 96

SRC_COMPONENT = "src"


def fail(msg: str) -> None:
    print(f"COVERAGE-PROD FAIL: {msg}", file=sys.stderr)
    sys.exit(1)


def mask_non_code(text: str) -> str:
    """Replaces every comment and the content of every string/char literal with
    spaces of equal length, preserving newlines (and therefore exact line numbers),
    so the brace counter in cfg_test_extents cannot be desynced by a brace character
    that happens to sit inside a string literal (this crate's own tests hold several
    r#"{"schemaName":...}"# JSON fixtures) or inside a comment. Adapted from
    scripts/gates/gate-mech1.py's strip_comments, which instead leaves string
    content intact; this checker additionally blanks it, because a brace inside a
    string would otherwise unbalance brace-depth counting rather than merely being
    counted as a `..` false positive.
    """

    def blank(s: str) -> str:
        return "".join(ch if ch == "\n" else " " for ch in s)

    out = []
    i, n = 0, len(text)
    while i < n:
        c = text[i]
        if c == "r":  # raw string literal
            m = re.match(r'r(#*)"', text[i:])
            if m:
                hashes = m.group(1)
                terminator = '"' + hashes
                start = i + m.end()
                end = text.find(terminator, start)
                end = n if end == -1 else end + len(terminator)
                out.append(blank(text[i:end]))
                i = end
                continue
        if c == '"':  # string literal
            j = i + 1
            while j < n:
                if text[j] == "\\" and j + 1 < n:
                    j += 2
                    continue
                if text[j] == '"':
                    j += 1
                    break
                j += 1
            out.append(blank(text[i:j]))
            i = j
            continue
        if c == "'" and i + 1 < n:  # char literal or lifetime
            m = re.match(r"'(?:\\.|[^\\'])'", text[i:])
            if m:
                out.append(blank(m.group(0)))
                i += len(m.group(0))
                continue
            out.append(c)
            i += 1
            continue
        if text.startswith("//", i):  # line comment
            j = text.find("\n", i)
            j = n if j == -1 else j
            out.append(blank(text[i:j]))
            i = j
            continue
        if text.startswith("/*", i):  # block comment, nesting
            depth, j = 1, i + 2
            while j < n and depth:
                if text.startswith("/*", j):
                    depth += 1
                    j += 2
                elif text.startswith("*/", j):
                    depth -= 1
                    j += 2
                else:
                    j += 1
            out.append(blank(text[i:j]))
            i = j
            continue
        out.append(c)
        i += 1
    return "".join(out)


CFG_TEST_RE = re.compile(r"#\s*\[\s*cfg\s*\(\s*test\s*\)\s*\]")


def lineno(text: str, pos: int) -> int:
    return text.count("\n", 0, pos) + 1


def cfg_test_extents(masked: str) -> list:
    """Returns one (start_line, end_line) 1-indexed inclusive range per `#[cfg(test)]`
    attribute found in `masked`, spanning from the attribute's own line through the
    end of the brace-delimited item it decorates, or through the terminating `;` for
    a bodyless item (a `#[cfg(test)] use ...;` or `#[cfg(test)] static ...;`)."""
    extents = []
    n = len(masked)
    for m in CFG_TEST_RE.finditer(masked):
        start_line = lineno(masked, m.start())
        i = m.end()
        depth = 0
        seen_brace = False
        end_pos = n - 1
        while i < n:
            c = masked[i]
            if c == "{":
                depth += 1
                seen_brace = True
            elif c == "}":
                depth -= 1
                if seen_brace and depth == 0:
                    end_pos = i
                    break
            elif c == ";" and not seen_brace and depth == 0:
                end_pos = i
                break
            i += 1
        extents.append((start_line, lineno(masked, end_pos)))
    return extents


def load_report(path: Path):
    if not path.is_file():
        fail(f"report file not found: {path}")
    try:
        raw = path.read_text()
    except OSError as e:
        fail(f"cannot read report {path}: {e}")
    try:
        return json.loads(raw)
    except json.JSONDecodeError as e:
        fail(f"cannot parse JSON report {path}: {e}")


def iter_report_files(data) -> list:
    files = []
    if not isinstance(data, dict):
        return files
    for entry in data.get("data", []) or []:
        if not isinstance(entry, dict):
            continue
        for f in entry.get("files", []) or []:
            if isinstance(f, dict):
                files.append(f)
    return files


def is_under_src(filename: str) -> bool:
    return SRC_COMPONENT in Path(filename).parts


# A small, regex-based extractor for exactly the two keys this check needs from each
# [[row]] block - see the module docstring's "gate-integrity task 6.4" note for why this
# is not a TOML parser.
ROW_BLOCK_RE = re.compile(r"\[\[row\]\](.*?)(?=\n\[\[row\]\]|\Z)", re.S)
CONDITION_RE = re.compile(r'condition\s*=\s*"((?:[^"\\]|\\.)*)"')
COVERS_ARRAY_RE = re.compile(r"covers\s*=\s*\[(.*?)\]", re.S)
COVERS_ITEM_RE = re.compile(r'"((?:[^"\\]|\\.)*)"')


def parse_covers_map(text: str) -> list:
    """Returns one (condition, [(path, first, last), ...]) pair per `[[row]]` block found
    in `text` - a row with no `covers` array, or an unparseable one, contributes an empty
    range list rather than being dropped, so the caller's own row/range count comparison
    stays meaningful."""
    rows = []
    for block in ROW_BLOCK_RE.findall(text):
        cond_m = CONDITION_RE.search(block)
        if not cond_m:
            continue
        condition = cond_m.group(1)
        entries = []
        covers_m = COVERS_ARRAY_RE.search(block)
        if covers_m:
            for item in COVERS_ITEM_RE.findall(covers_m.group(1)):
                path, sep, rng = item.rpartition(":")
                if not sep or "-" not in rng:
                    continue
                first_s, _, last_s = rng.partition("-")
                try:
                    entries.append((path, int(first_s), int(last_s)))
                except ValueError:
                    continue
        rows.append((condition, entries))
    return rows


def build_line_counts(report_data) -> dict:
    """Per-file `{line: max(count)}`, over every `hasCount` segment in every file the
    report names - not only the ones under `src/`, since a `covers` fixture in
    `tests/fixtures/coverage/` legitimately points elsewhere."""
    line_counts = {}
    for f in iter_report_files(report_data):
        filename = f.get("filename")
        if not isinstance(filename, str):
            continue
        counts = {}
        for seg in f.get("segments") or []:
            if not isinstance(seg, (list, tuple)) or len(seg) < 4:
                continue
            line, _col, count, has_count = seg[0], seg[1], seg[2], seg[3]
            if not has_count:
                continue
            prev = counts.get(line)
            if prev is None or count > prev:
                counts[line] = count
        line_counts[filename] = counts
    return line_counts


def find_counts_for_path(line_counts: dict, path: str):
    """Match a `covers` entry's `path` against the report's own filenames - exactly, or as
    a `/`-bounded suffix, since a real `cargo llvm-cov` report names files by an absolute
    path (see `classify_file`'s own resolution) while a fixture report may name one
    relatively already."""
    for filename, counts in line_counts.items():
        if filename == path or filename.endswith("/" + path):
            return counts
    return None


def check_covers_ranges(report_data, toml_path: Path) -> None:
    """gate-integrity task 6.4: every line of every `covers` range in the degraded-coverage
    map SHALL be executed. Fails naming the offending row's `condition`, per the spec, and
    cannot pass vacuously: an unreadable map, a map with no rows, an entry whose every
    `covers` array is empty, or a map holding fewer total ranges than rows are each a
    failure, not a report of 100% over nothing."""
    if not toml_path.is_file():
        fail(f"degraded-coverage map not found: {toml_path}")
    text = toml_path.read_text()

    total_row_blocks = text.count("[[row]]")
    if total_row_blocks == 0:
        fail(f"{toml_path} names no [[row]] entries - refusing to report every range covered")

    rows = parse_covers_map(text)
    total_ranges = sum(len(entries) for _, entries in rows)
    if total_ranges == 0:
        fail(
            f"{toml_path}: every row's \"covers\" array is empty - refusing to report "
            "every range covered"
        )
    if total_ranges < total_row_blocks:
        fail(
            f"{toml_path} holds {total_row_blocks} row(s) but only {total_ranges} "
            "covers range(s) in total - dropping a range is a failure, not a silent "
            "reduction in what is proved"
        )

    line_counts = build_line_counts(report_data)
    failures = []
    for condition, entries in rows:
        for path, first, last in entries:
            counts = find_counts_for_path(line_counts, path)
            if counts is None:
                failures.append(f"{condition!r}: {path}:{first}-{last} - report names no such file")
                continue
            instrumented = [ln for ln in range(first, last + 1) if ln in counts]
            if not instrumented:
                failures.append(
                    f"{condition!r}: {path}:{first}-{last} - report instruments no line "
                    "in this range"
                )
                continue
            cold = [ln for ln in instrumented if counts[ln] == 0]
            if cold:
                failures.append(
                    f"{condition!r}: {path}:{first}-{last} has uncovered line(s) {cold}"
                )
    if failures:
        fail("degraded-coverage covers range(s) uncovered:\n  " + "\n  ".join(failures))

    print(
        f"COVERAGE-PROD DEGRADED-COVERS OK: {total_row_blocks} row(s), {total_ranges} "
        f"range(s) from {toml_path}, all covered"
    )


def classify_file(filename: str, segments: list) -> dict:
    src_path = Path(filename)
    if not src_path.is_file():
        fail(
            f"cannot locate source file for report entry {filename!r} "
            f"(expected it at that path, relative to the working directory)"
        )
    text = src_path.read_text()
    masked = mask_non_code(text)
    extents = cfg_test_extents(masked)

    test_lines = set()
    for start, end in extents:
        test_lines.update(range(start, end + 1))

    line_count = {}
    for seg in segments:
        if not isinstance(seg, (list, tuple)) or len(seg) < 4:
            continue
        line, _col, count, has_count = seg[0], seg[1], seg[2], seg[3]
        if not has_count:
            continue
        prev = line_count.get(line)
        if prev is None or count > prev:
            line_count[line] = count

    prod_instr = prod_cov = test_instr = test_cov = 0
    for line, count in line_count.items():
        if line in test_lines:
            test_instr += 1
            if count > 0:
                test_cov += 1
        else:
            prod_instr += 1
            if count > 0:
                prod_cov += 1

    return {
        "extents": len(extents),
        "prod_instr": prod_instr,
        "prod_cov": prod_cov,
        "test_instr": test_instr,
        "test_cov": test_cov,
    }


def main(argv: list) -> int:
    if len(argv) < 2:
        fail("usage: python3 coverage-prod.py <llvm-cov-json-report> [<degraded-coverage-map>]")
    report_path = Path(argv[1])
    covers_map_path = Path(argv[2]) if len(argv) > 2 else None

    floor = DEFAULT_PROD_MIN
    override = os.environ.get("PROD_MIN")
    if override:
        try:
            floor = float(override)
        except ValueError:
            fail(f"PROD_MIN={override!r} is not a number")

    data = load_report(report_path)

    if covers_map_path is not None:
        check_covers_ranges(data, covers_map_path)

    all_files = iter_report_files(data)
    src_files = [
        f
        for f in all_files
        if isinstance(f.get("filename"), str) and is_under_src(f["filename"])
    ]

    if not src_files:
        fail(
            f"report names no file under {SRC_COMPONENT!r}/ "
            f"({len(all_files)} file(s) total) - refusing to report 100% of nothing"
        )

    totals = {"prod_instr": 0, "prod_cov": 0, "test_instr": 0, "test_cov": 0}
    per_file = []
    for f in src_files:
        filename = f["filename"]
        result = classify_file(filename, f.get("segments") or [])
        per_file.append((filename, result))
        for k in totals:
            totals[k] += result[k]
        print(
            f"COVERAGE-PROD FILE {filename} extents={result['extents']} "
            f"prod_instrumented={result['prod_instr']} prod_covered={result['prod_cov']} "
            f"test_instrumented={result['test_instr']} test_covered={result['test_cov']}"
        )

    if totals["prod_instr"] == 0:
        fail(
            "report names 0 production-instrumented lines under "
            f"{SRC_COMPONENT}/ - refusing to report 100% of nothing"
        )

    prod_pct = 100.0 * totals["prod_cov"] / totals["prod_instr"]
    test_pct = (
        100.0 * totals["test_cov"] / totals["test_instr"] if totals["test_instr"] else 0.0
    )

    if prod_pct < floor:
        worst = sorted(
            per_file,
            key=lambda kv: kv[1]["prod_instr"] - kv[1]["prod_cov"],
            reverse=True,
        )
        worst = [
            f"{name} ({r['prod_instr'] - r['prod_cov']})"
            for name, r in worst
            if r["prod_instr"] - r["prod_cov"] > 0
        ][:5]
        fail(
            f"production coverage {prod_pct:.2f}% ({totals['prod_cov']}/{totals['prod_instr']}) "
            f"< floor {floor}% (line hasCount/max-count rule); "
            f"most uncovered production lines: {', '.join(worst) or 'n/a'}"
        )

    print(
        f"COVERAGE-PROD OK: production {prod_pct:.2f}% "
        f"({totals['prod_cov']}/{totals['prod_instr']}) >= floor {floor}%; "
        f"test-module {test_pct:.2f}% ({totals['test_cov']}/{totals['test_instr']}); "
        f"{len(src_files)} file(s) under {SRC_COMPONENT}/ (line hasCount/max-count rule)"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: tests/coverage_prod.rs
//! Drives `scripts/coverage-prod.py` — the production-slice coverage floor — against
//! checked-in fixture reports under `tests/fixtures/coverage/`, and against synthetic
//! reports built at test time over this crate's own real `src/*.rs` files.
//!
//! See `openspec/changes/gate-integrity/design.md` -> Decision 1/1a and the
//! `quality-gates` spec's "The coverage floor is measured against production code" for the
//! contract this file exercises. RED at task 1.1 time: `scripts/coverage-prod.py` does not
//! exist yet.

use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::sync::atomic::{AtomicUsize, Ordering};

fn manifest_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
}

fn script_path() -> PathBuf {
    manifest_dir().join("scripts/coverage-prod.py")
}

fn fixtures_dir() -> PathBuf {
    manifest_dir().join("tests/fixtures/coverage")
}

/// Runs the checker against `report`, with an optional `PROD_MIN` override. Returns
/// whether it exited 0 and the combined stdout+stderr text.
fn run_checker(report: &Path, prod_min: Option<&str>) -> (bool, String) {
    let mut cmd = Command::new("python3");
    cmd.arg(script_path()).arg(report).current_dir(manifest_dir());
    if let Some(min) = prod_min {
        cmd.env("PROD_MIN", min);
    }
    let output = cmd
        .output()
        .unwrap_or_else(|e| panic!("failed to spawn python3 {}: {e}", script_path().display()));
    let mut combined = String::from_utf8_lossy(&output.stdout).into_owned();
    combined.push_str(&String::from_utf8_lossy(&output.stderr));
    (output.status.success(), combined)
}

fn fixture(name: &str) -> PathBuf {
    fixtures_dir().join(name)
}

#[test]
fn healthy_report_passes() {
    let (ok, out) = run_checker(&fixture("healthy.json"), None);
    assert!(ok, "expected the healthy fixture to pass:\n{out}");
    assert!(
        out.contains("COVERAGE-PROD OK"),
        "expected an OK line in output:\n{out}"
    );
}

#[test]
fn one_module_zeroed_fails() {
    let (ok, out) = run_checker(&fixture("one_module_zeroed.json"), None);
    assert!(
        !ok,
        "expected the fixture with production lines zeroed to fail:\n{out}"
    );
}

#[test]
fn empty_report_fails() {
    let (ok, out) = run_checker(&fixture("empty.json"), None);
    assert!(
        !ok,
        "expected an empty report (no files at all) to fail rather than report 100% of nothing:\n{out}"
    );
}

#[test]
fn src_less_report_fails() {
    let (ok, out) = run_checker(&fixture("no_src.json"), None);
    assert!(
        !ok,
        "expected a report naming no file under src/ to fail:\n{out}"
    );
}

#[test]
fn malformed_report_fails() {
    let (ok, out) = run_checker(&fixture("malformed.json"), None);
    assert!(!ok, "expected an unparseable report to fail:\n{out}");
}

#[test]
fn absent_report_fails() {
    let (ok, out) = run_checker(&fixture("does-not-exist.json"), None);
    assert!(!ok, "expected an absent report path to fail:\n{out}");
}

#[test]
fn zero_production_lines_fails() {
    let (ok, out) = run_checker(&fixture("zero_prod_lines.json"), None);
    assert!(
        !ok,
        "expected a report whose production-line count is zero to fail:\n{out}"
    );
}

/// Extracts an integer following `key=` on a line naming `filename`, from the checker's
/// `COVERAGE-PROD FILE <filename> extents=N prod_instrumented=N ...` output. Panics with
/// the full output if the file's line, or the key on it, cannot be found — a missing line
/// is a harness problem, not a passing/failing result to fall back on silently.
fn extract_metric(out: &str, filename: &str, key: &str) -> u64 {
    let prefix = format!("{key}=");
    for line in out.lines() {
        if !line.contains(filename) {
            continue;
        }
        for token in line.split_whitespace() {
            if let Some(value) = token.strip_prefix(&prefix) {
                return value
                    .parse()
                    .unwrap_or_else(|e| panic!("bad {key} value {value:?} in line {line:?}: {e}"));
            }
        }
    }
    panic!("no {key:?} found for {filename:?} in output:\n{out}");
}

/// A scratch temp file removed on drop, used for synthetic reports generated at test time
/// over this crate's own real `src/*.rs` files — those are too large to check in as fixed
/// fixtures, so this file builds them on the fly instead.
struct ScratchFile {
    path: PathBuf,
}

impl ScratchFile {
    fn new(label: &str, contents: &str) -> Self {
        static COUNTER: AtomicUsize = AtomicUsize::new(0);
        let n = COUNTER.fetch_add(1, Ordering::Relaxed);
        let path = std::env::temp_dir().join(format!(
            "herdr-openspec-coverage-prod-{}-{n}-{label}.json",
            std::process::id()
        ));
        fs::write(&path, contents).expect("write scratch report");
        Self { path }
    }

    fn path(&self) -> &Path {
        &self.path
    }
}

impl Drop for ScratchFile {
    fn drop(&mut self) {
        let _ = fs::remove_file(&self.path);
    }
}

/// Builds a synthetic llvm-cov JSON export naming a single real file (`rel_path`, relative
/// to the repository root) with every one of its lines marked instrumented and covered —
/// enough to exercise the classifier's per-file extent tracking without needing a real
/// `cargo llvm-cov` run.
fn synthetic_all_covered_report(rel_path: &str) -> String {
    let text = fs::read_to_string(manifest_dir().join(rel_path))
        .unwrap_or_else(|e| panic!("read {rel_path} to build synthetic report: {e}"));
    let n = text.lines().count();
    let mut segments = Vec::with_capacity(n);
    for line in 1..=n {
        segments.push(format!("[{line},1,1,true,true,false]"));
    }
    format!(
        r#"{{"data":[{{"files":[{{"filename":"{rel_path}","segments":[{}]}}]}}]}}"#,
        segments.join(",")
    )
}

/// Task 1.2b: the naive "first line-anchored `#[cfg(test)]`" cut would classify
/// `src/changes.rs` as having only 76 production lines (everything from line 77 on is
/// `mod conformance`, a `#[cfg(test)]` item, and the naive cut treats the rest of the file
/// as test too). The correct brace-extent classifier must instead exclude only the actual
/// `#[cfg(test)]` items' extents and count everything else — thousands of lines — as
/// production.
#[test]
fn changes_rs_contributes_its_real_production_body_not_76_lines() {
    let report = synthetic_all_covered_report("src/changes.rs");
    let scratch = ScratchFile::new("changes-rs", &report);
    let (ok, out) = run_checker(scratch.path(), Some("0"));
    assert!(ok, "synthetic all-covered report should pass:\n{out}");

    let extents = extract_metric(&out, "src/changes.rs", "extents");
    let prod_instr = extract_metric(&out, "src/changes.rs", "prod_instrumented");

    assert!(
        extents >= 3,
        "expected at least 3 #[cfg(test)] extents in src/changes.rs, found {extents}:\n{out}"
    );
    assert!(
        prod_instr > 1000,
        "expected src/changes.rs's production line count to be its real (large) body, \
         not the naive first-occurrence cut's 76 lines; got {prod_instr}:\n{out}"
    );
}

#[test]
fn cli_rs_contributes_more_than_the_naive_cut() {
    let report = synthetic_all_covered_report("src/cli.rs");
    let scratch = ScratchFile::new("cli-rs", &report);
    let (ok, out) = run_checker(scratch.path(), Some("0"));
    assert!(ok, "synthetic all-covered report should pass:\n{out}");

    let extents = extract_metric(&out, "src/cli.rs", "extents");
    let prod_instr = extract_metric(&out, "src/cli.rs", "prod_instrumented");

    assert!(
        extents >= 10,
        "expected at least 10 #[cfg(test)] extents in src/cli.rs, found {extents}:\n{out}"
    );
    // The naive cut takes everything before line 317 as production (~300 lines) and
    // discards the rest; the real production body extends past the test modules too.
    assert!(
        prod_instr > 300,
        "expected src/cli.rs's production line count to exceed the naive cut's ~300 lines; \
         got {prod_instr}:\n{out}"
    );
}

#[test]
fn lib_rs_contributes_more_than_the_naive_cut() {
    let report = synthetic_all_covered_report("src/lib.rs");
    let scratch = ScratchFile::new("lib-rs", &report);
    let (ok, out) = run_checker(scratch.path(), Some("0"));
    assert!(ok, "synthetic all-covered report should pass:\n{out}");

    let extents = extract_metric(&out, "src/lib.rs", "extents");
    let prod_instr = extract_metric(&out, "src/lib.rs", "prod_instrumented");

    assert!(
        extents >= 2,
        "expected at least 2 #[cfg(test)] extents in src/lib.rs, found {extents}:\n{out}"
    );
    // The naive cut takes everything before line 37 (`mod testutil`) as production; the
    // real body includes every production module declared after it (`pub mod changes`,
    // `pub mod cli`, ... at lines 9-20, all before line 37, plus `pid()` at line ~30).
    // The point of this test is that the classifier does not stop at the naive cut's
    // ~36 lines once a later, larger `#[cfg(test)]` module (line 987) is also present.
    assert!(
        prod_instr > 36,
        "expected src/lib.rs's production line count to exceed the naive cut's ~36 lines; \
         got {prod_instr}:\n{out}"
    );
}

/// Task 1.2b's negative control: move a production function into a `#[cfg(test)]` module
/// (here, `tests/fixtures/coverage/src/sample_moved.rs` — `sample.rs` with `#[cfg(test)]`
/// added directly above `pub fn sub`) and require the reported production line count to
/// fall, using checked-in fixture reports that mark the same two candidate lines in each
/// file (shifted by the one inserted attribute line).
#[test]
fn moving_a_production_fn_into_cfg_test_lowers_the_count() {
    let (before_ok, before_out) = run_checker(&fixture("move_control_before.json"), Some("0"));
    assert!(before_ok, "before-move report should pass:\n{before_out}");
    let before = extract_metric(
        &before_out,
        "tests/fixtures/coverage/src/sample.rs",
        "prod_instrumented",
    );

    let (after_ok, after_out) = run_checker(&fixture("move_control_after.json"), Some("0"));
    assert!(after_ok, "after-move report should pass:\n{after_out}");
    let after = extract_metric(
        &after_out,
        "tests/fixtures/coverage/src/sample_moved.rs",
        "prod_instrumented",
    );

    assert_eq!(before, 2, "sanity: both candidate lines start out production:\n{before_out}");
    assert!(
        after < before,
        "moving a production fn into #[cfg(test)] must lower the production line count: \
         before={before} after={after}\n{before_out}\n{after_out}"
    );
}

/// The floor is exceedable: running with `PROD_MIN` far above the healthy fixture's
/// figure must fail, proving the threshold is enforced rather than merely printed.
#[test]
fn raising_the_floor_above_measurement_fails() {
    let (ok, out) = run_checker(&fixture("healthy.json"), Some("100.5"));
    assert!(
        !ok,
        "expected a floor above the measured figure to fail:\n{out}"
    );
}
